introduce typed configuration API

eiennohito opened the pull request:
Fixes #165, adds checks for #157 

Changes:
* Split Settings into three parts:
  * Settings - they become untyped configuration. It does relative path resolution.
  * Config - typed configuration that should be used as a general API. It uses only resolved paths and does not perform any path resolution.
  * SettingsAnchor - path resolution logic, most of the type the explicit usage is not needed, but in complex cases it is available as public API.
* Removed FormatterPlugin from configuration. CommandLine uses hardcoded implementations. Reasoning: it was not used and providing this logic as a plugin has almost no merits. Custom usages won't use it and the current usage was not really customizable.
* Added `--systemDict` and `--userDict` options to `SudachiCommandLine`, now specifying dictionaries became easy as it should have been!
* Slightly refactored logging configuration.

I've tried to write down the usage and assumptions in JavaDoc for everything.

Sonarcloud fails because I have modified logging configuration in `SudachiCommandLine`

eiennohito committed:
introduce typed configuration API

eiennohito committed:
add test for fromClasspathMerged

eiennohito committed:
improve ergonomics for settings anchors

eiennohito committed:
support loading byte buffer resources from classpath

eiennohito committed:
redo anchor chaining

eiennohito committed:
add javadoc for Config

eiennohito committed:
add javadoc for Settings

eiennohito committed:
add javadoc for anchor

eiennohito committed:
rename Settings.merge -> read for such operations

also, deprecate getPath

kazuma-t commented on src/main/java/com/worksap/nlp/sudachi/dictionary/BinaryDictionary.java:
```diff
@@ -55,23 +56,39 @@ public BinaryDictionary(ByteBuffer dictionary) throws IOException {
+    }
+
+    public static BinaryDictionary readUserDictionary(String fileName) throws IOException {
+        return loadSystem(MMap.map(fileName));
```
`loadUser()` ?

eiennohito replied:
`loadUser`/`loadSystem` for both of them then

kazuma-t replied:
Doesn't `loadSystem(MMap.map(fileName))` have to be `loadUser(MMap.map(fileName))`?

eiennohito replied:
Certainly it should be that way.

kazuma-t commented on src/main/java/com/worksap/nlp/sudachi/Settings.java:
```diff
@@ -73,41 +79,134 @@
                 JsonObject root = (JsonObject) rootStr;
-                String basePath = root.getString("path", path);
-                return new Settings(root, basePath);
+                String basePath = root.getString("path", null);
```
The priority of this part is the same as before the improvement.

config file > path parameter (settings anchor)

I think it would be better to reverse it. Do you want to discuss this in another issue?

eiennohito replied:
I thought that the in-config value should have higher priority than out-of-band configuration, so kept the current behavior, but I can change it. With the current structure of anchors, special treatment of the `path` key probably can be completely removed, but I am not sure because I am still not fully aware of all its usages.

kazuma-t replied:
OK. Let's discuss this separately.

kazuma-t commented on src/main/java/com/worksap/nlp/sudachi/Settings.java:
```diff
@@ -393,42 +605,18 @@ void merge(Settings settings) {
-        Map<Integer, JsonValue> replaceItems = new HashMap<>();
-
-        for (JsonValue item : (JsonArray) value) {
-            boolean isReplaced = false;
```
If we eliminate this part, users will not be able to partially rewrite the plugin settings.
Is it your decision that this behavior is confusing and should not be removed?

eiennohito replied:
I would like to move smart merging to `Config`s as much as possible and make Settings merging to be a simple JSON merging. Plugin-aware merging definitely should go there.

kazuma-t replied:
I see.

kazuma-t commented on src/main/java/com/worksap/nlp/sudachi/SettingsAnchor.java:
```diff
@@ -0,0 +1,349 @@
+            for (SettingsAnchor item : items) {
+                if (item instanceof Chain) {
+                    Chain c = (Chain) item;
+                    children.forEach(this::add);
```
`c.children.forEach(this::add)` ?

eiennohito replied:
Nice catch!

kazuma-t commented on src/main/java/com/worksap/nlp/sudachi/SettingsAnchor.java:
```diff
@@ -0,0 +1,349 @@
+            if (equals(other)) {
+                return this;
+            }
+            children.forEach(this::add);
```
`other.children` ?

eiennohito replied:
Nice catch!

kazuma-t commented on src/main/java/com/worksap/nlp/sudachi/SudachiCommandLine.java:
```diff
@@ -87,25 +92,24 @@ static void run(Tokenizer tokenizer, Tokenizer.SplitMode mode, InputStream input
             }
         } else {
-            formatter = formatters.get(0);
+            formatter = new SimpleMorphemeFormatter();
```
Since SimpleMorphemeFormatter is used as the definitive formatter, other formatters cannot be used.
https://github.com/WorksApplications/chikkar#sudachi-%E3%82%B3%E3%83%9E%E3%83%B3%E3%83%89%E3%83%A9%E3%82%A4%E3%83%B3%E3%83%84%E3%83%BC%E3%83%AB%E3%81%A7%E3%81%AE%E5%88%A9%E7%94%A8-using-chikkar-with-the-sudachi-cli

eiennohito replied:
Mixing analysis config and morpheme formatting does not seem good for me.

Proposal: introduce `--format` flag which will accept some basic formatter types or Java class name, which will be instantiated to do the formatting.

kazuma-t replied:
OK. I don't want Chikkar to stop working, so please do something about it.

eiennohito replied:
Added `--format` parameter which now accepts full class name of the formatter class. Probably it can be extended to support short aliases for current formatters as well.

kazuma-t commented on src/test/java/com/worksap/nlp/sudachi/SettingsTest.java:
```diff
@@ -35,89 +30,89 @@
     @Test
     public void getIntWithDefaultValue() {
-        Settings settings = Settings.parseSettings(null, "{\"foo\":123}");
+        Settings settings = Settings.parse("{\"foo\":123}", null);
```
It's better to unify with `SettingsAnchor.none()`.

eiennohito replied:
I wanted to leave some tests which explicitly use null as the anchor parameter.

kazuma-t replied:
I understand. But I think it's better to add a comment.

eiennohito committed:
apply review comments

eiennohito committed:
add format argument to SudachiCommandLine

It accepts arbitrary formatter class name.
Additional settings passed as json string are passed to all
created formatters now.

eiennohito committed:
check for invalid classes as well by capturing ClassCastException

eiennohito committed:
loadSystem -> loadUser

eiennohito committed:
add comments on nulls in SettingsTest
